node/heartbeat/monitor.go

```diff
@@ -247,12 +252,27 @@ func (m *Monitor) ProcessReceivedMessage(message p2p.MessageP2P, fromConnectedPe
 		return err
 	}
 
-	isHardforkTrigger := bytes.Equal(hbRecv.Payload, []byte(hardforkTriggerString))
+	isHardforkTrigger, err := m.hardforkTrigger.TriggerReceived(message.Data(), hbRecv.Payload, hbRecv.Pubkey)
 	if isHardforkTrigger {
-		errHardforkTrigger := m.hardforkTrigger.TriggerReceived(hbRecv.Pubkey, hbRecv.Payload)
-		if errHardforkTrigger != nil {
-			return errHardforkTrigger
-		}
+		return err
+	}
+
+	if !bytes.Equal(hbRecv.Pid, message.Peer().Bytes()) {
```

Comment:
the hbRecv is created from the message, and the pid is set to the Peer.Bytes() if the pid does not exist in a map. When can they be different?

Reply:
The whole idea of this fix was to include the peer id (pid) inside the message that will get signed. If a node decides to "replay" or re-broadcast a message containing a pid other then the self, it will immediately be blacklisted by other peer. This is done as to prevent some malicious nodes to replay other node's heartbeat message in order to confuse and disrupt the network sharding's data collector.